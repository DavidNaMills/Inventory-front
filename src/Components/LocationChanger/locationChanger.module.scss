.locationChanger{
    display: flex;
    align-items: center;

    &__selectStyle{
        border-radius: 10px;
        width: 200px;
        padding: 3px;
    }

    &__bold{
        font-weight: bold;
        margin-right: 15px;
    }
}