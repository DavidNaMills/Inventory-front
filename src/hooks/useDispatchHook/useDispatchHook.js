import { useDispatch } from 'react-redux';

import { userLogin, userLogout, changeCurrentLocation } from '../../store/userStore/userActions';
import { startInitTypes, addNewProductType, addNewLocation, removeLocationType } from '../../store/typesStore/typesStoreActions';
import {createMessage} from '../../store/messagingStore/messagingActions';

const useDispatchHook = () => {
    const dispatch = useDispatch();

    const userLoginDispatch = (userData) => dispatch(userLogin(userData));
    const userLogoutDispatch = () => dispatch(userLogout());
    const changeCurrentLocationDispatch = (id) => dispatch(changeCurrentLocation(id));

    const initTypesDispatch = () => dispatch(startInitTypes());
    const addNewProductDispatch = (newType) => dispatch(addNewProductType(newType));
    const addNewLocationDispatch = (newLocation) => dispatch(addNewLocation(newLocation));
    const removeLocationDispatch = (id) => dispatch(removeLocationType(id));

    const createMessageDispatch = (data) => dispatch(createMessage(data));

    return {
        userLoginDispatch,
        userLogoutDispatch,
        changeCurrentLocationDispatch,

        initTypesDispatch,
        addNewProductDispatch,
        addNewLocationDispatch,
        removeLocationDispatch,

        createMessageDispatch
    }
}

export default useDispatchHook;