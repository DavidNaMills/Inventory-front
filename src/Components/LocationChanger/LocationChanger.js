import React from 'react';
import classes from './locationChanger.module.scss';
import { useSelector } from 'react-redux';
import useDispatchHook from '../../hooks/useDispatchHook/useDispatchHook';
import PlaceIcon from '@material-ui/icons/Place';


const LocationChanger = () => {
    const { staff, types } = useSelector(state => state);
    const { changeCurrentLocationDispatch } = useDispatchHook();
    return (
        (staff.token && Object.keys(types.locations).length > 0)
        && <div className={classes.locationChanger}>
            <span className={classes.locationChanger__bold}><PlaceIcon/> </span>


            {(staff.staff.baseId.length > 1)
                ?
                <select onChange={(e) => changeCurrentLocationDispatch(e.target.value)} className={classes.locationChanger__selectStyle}>
                    {staff.staff.baseId.map(x => <option key={x} value={x}>{types.locations[x].displayValue}</option>)}
                </select>
                : <p>{types.locations[staff.staff.baseId[0]].displayValue}</p>
            }
        </div>

    )
}
export default React.memo(LocationChanger);