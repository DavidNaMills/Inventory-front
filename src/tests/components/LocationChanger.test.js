import React from 'react';
import '../testHelpers/context';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

jest.mock('react-redux');
import { useSelector } from 'react-redux';

jest.mock('../../hooks/useDispatchHook/useDispatchHook');
import useDispatchHook from '../../hooks/useDispatchHook/useDispatchHook';

import LocationChanger from '../../Components/LocationChanger/LocationChanger';
import PlaceIcon from '@material-ui/icons/Place';


const mockStaff = {
    token: 'fdsfdfdfdsfsa',
    staff: {
        name: 'David',
        baseId: ['123', '456123']
    }
};

const mockStaffSingle = {
    token: 'fdsfdfdfdsfsa',
    staff: {
        name: 'David',
        baseId: ['123']
    }
};

const mockLocations = {
    locations: {
        '456123': { displayValue: 'home2' },
        '123': { displayValue: 'home' }
    }
}


afterEach(() => {
    jest.clearAllMocks();
})

const mockDispatch = jest.fn();
beforeEach(()=>{
    useDispatchHook.mockImplementation(()=>({
        changeCurrentLocationDispatch: mockDispatch
    }));
})

describe('<LocationChanger /> test suite', () => {
    describe('basic tests', () => {
        beforeEach(() => {
            useSelector.mockImplementation(() => ({
                staff: mockStaff,
                types: mockLocations
            }))
        });

        it('matches snapshot', () => {
            const comp = renderer.create(<LocationChanger />);
            const app = comp.toJSON();
            expect(app).toMatchSnapshot();
        });

        it('exists', () => {
            const wrapper = mount(<LocationChanger />);
            expect(wrapper.exists).toBeTruthy();
        });
    });

    describe('Locations are greater than 1 and staff.token present', () => {
        let wrapper;

        beforeEach(()=>{
            useSelector.mockImplementation(() => ({
                staff: mockStaff,
                types: mockLocations
            }));
            wrapper = mount(<LocationChanger />);
        });

        it('renders a <div> with className="locationChanger" a <span/> with className="locationChanger__bold" and <PlaceIcon/>', () => { 
            const div = wrapper.find('div');
            const span = wrapper.find('span');
            expect(div.length).toBe(1);
            expect(wrapper.find(PlaceIcon).length).toBe(1);
            expect(span.exists).toBeTruthy();
            expect(span.props()).toHaveProperty('className', 'locationChanger__bold');
            expect(div.props()).toHaveProperty('className', 'locationChanger');
        });
        
        it('renders a <select/> element if staff.baseId[] length is greater than 1', () => { 
            const select = wrapper.find('select');
            expect(select.length).toBe(1);
            expect(select.props().children.length).toBe(2);
            expect(select.props()).toHaveProperty('className', 'locationChanger__selectStyle');
        });
        
        it('fires the changeCurrentLocationDispatch method on select change', () => { 
            wrapper.find('select').simulate('change');
            expect(mockDispatch).toHaveBeenCalledTimes(1);
            expect(mockDispatch).toHaveBeenCalledWith('123');
        });
    });

    it('renders the location name if staff.baseId[] length is only 1 ', () => { 
        useSelector.mockImplementation(() => ({
            staff: mockStaffSingle,
            types: mockLocations
        }));
        const wrapper = shallow(<LocationChanger />);
        expect(wrapper.find('.locationChanger__selectStyle').length).toBe(0);
        expect(wrapper.find('p').length).toBe(1);
    });

    it('returns null if no staff.token present but types.locations are present', () => {
        useSelector.mockImplementation(() => ({
            staff: {token: null},
            types: mockLocations
        }));
        const wrapper = shallow(<LocationChanger />);
        expect(wrapper.find('.locationChanger__selectStyle').length).toBe(0);
        expect(wrapper.find('p').length).toBe(0);
     });

     it('returns null if types.locations length is 0 and staff.token exists', () => {
        useSelector.mockImplementation(() => ({
            staff: {token: 'fsdfda'},
            types: {locations: {}}
        }));
        const wrapper = shallow(<LocationChanger />);
        expect(wrapper.find('.locationChanger__selectStyle').length).toBe(0);
        expect(wrapper.find('p').length).toBe(0);
     });
});