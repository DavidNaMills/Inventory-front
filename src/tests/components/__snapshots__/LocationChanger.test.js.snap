// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationChanger /> test suite basic tests matches snapshot 1`] = `
<div
  className="locationChanger"
>
  <span
    className="locationChanger__bold"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
      />
    </svg>
     
  </span>
  <select
    className="locationChanger__selectStyle"
    onChange={[Function]}
  >
    <option
      value="123"
    >
      home
    </option>
    <option
      value="456123"
    >
      home2
    </option>
  </select>
</div>
`;
